### ISSUE TITLE
Unable to do tensor comparison on Metal Performance Shaders (MPS)

### ISSUE THREAD

**Issue by jublikon** [pytorch/pytorch#101566]
### 🐛 Describe the bug

Using the official [PyTorch example for MNIST](https://github.com/pytorch/examples/blob/main/mnist/main.py), 
I am unable to compute a[ tensor element-wise equality](https://pytorch.org/docs/stable/generated/torch.eq.html) on the [MPS backend](https://developer.apple.com/metal/pytorch/), supported by Apple Metal.
The problem happens in line 64, where the amount of correct predictions is computed: 

`correct += pred.eq(target.view_as(pred)).sum().item()`

So, the accuracy appears always to be 0, in line 68 to 70:

```
 print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))
```
That is, because the eq comparison evaluates everything to `False`. 

However, when I turn MPS off, by switching the `--no-mps` flag to True, I do obtain accuracies and the network trains as expected on CPU.

Why is that the case? Am I missing here something within my configuration or is this a bug?
I would really love to use the AMD GPU of my Intel-powered MacBook Pro.

### Versions

PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.26.3
Libc version: N/A

Python version: 3.10.11 | packaged by conda-forge | (main, May 10 2023, 19:07:22) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.3.1-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.0
[pip3] torchaudio==2.0.0
[pip3] torchvision==0.15.1a0
[conda] mkl                       2022.2.1         h44ed08c_16952    conda-forge
[conda] numpy                     1.24.3          py310h7451ae0_0    conda-forge
[conda] pytorch                   2.0.0           cpu_py310h2bbf33f_0    conda-forge
[conda] torchaudio                2.0.0                 py310_cpu    pytorch
[conda] torchvision               0.15.1          cpu_py310hfdc6817_0    conda-forge

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by spzala** [comment]
With the current `main` (`version 2.2.0`), I couldn't reproduce issue on mps. I get expected accuracy `Test set: Average loss: 0.0276, Accuracy: 9917/10000 (99%)` I also see expected result of this small snippet on eq comparison on mps device: 
```
import torch
cuda = torch.cuda.is_available()
mps = torch.backends.mps.is_available()
if cuda:
    device = torch.device("cuda")
elif mps:
    device = torch.device("mps")
else:
    device = torch.device("cpu")
print(torch.eq(torch.tensor([[1, 2], [3, 4]]), torch.tensor([[1, 1], [4, 4]])))
```

We should close this issue if you agree it's not an issue any more. @jublikon cc @albanD Thanks!

**Comment by albanD** [comment]
Thanks for checking. Yes we should close this if it is indeed fixed in main
